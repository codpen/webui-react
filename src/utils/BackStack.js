import Stack from './Stack'
// A very simple stack implementation to handle back links in remote explorer
class BackStack {
    constructor() {
        this.backStack = new Stack();
        this.forwardStack = new Stack();
    }

    getLength() {
        return this.backStack.getLength();
    }

    push(item) {

        this.backStack.push(item);
        this.forwardStack.empty();

    }

    pop() {
        const temp = this.backStack.pop();
        return temp;
    }

    peek() {
        return this.backStack.peek();
    }

    empty() {
        this.backStack.empty();
        this.forwardStack.empty();
    }

    moveBack() {
        const temp = this.backStack.moveBack();
        if(temp)
            this.forwardStack.push(temp);
        return temp;
    }

    moveForward(){
        const temp = this.forwardStack.pop();
        if(temp){
            //Pop was successful
            this.backStack.push(temp);
        }
        return temp;

    }



}

export default BackStack;