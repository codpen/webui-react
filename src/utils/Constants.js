/**
 * Returns a string array with possible filtering options.
 * @type {string[]}
 */
export const visibilityFilteringOptions = ["Images", "Pdf", "Videos"];

/**
 * Refresh timeout for checking backend status.
 * In milliseconds.
 * @type {number}
 */
export const STATUS_REFRESH_TIMEOUT = 1000 * 5;
/**
 * Refresh timeout for checking if the config is created.
 * In milliseconds.
 * @type {number}
 */
export const NEW_DRIVE_CONFIG_REFRESH_TIMEOUT = 1000;
export const MODAL_ROOT_ELEMENT = "modal-root";

export const USER_NAME_KEY = "userName";
export const PASSWORD_KEY = "password";
export const IP_ADDRESS_KEY = "ipAddress";

export const FILES_VIEW_HEIGHT = "75vh";