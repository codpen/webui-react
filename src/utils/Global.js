global.ipAddress = "http://127.0.0.1:5572";