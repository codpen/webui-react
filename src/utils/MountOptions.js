export const vfsOptions = {
    "CacheMaxAge": {
        Type: 'int',
        Default: 3600000000000
    },
    "CacheMaxSize": {
        Type: 'int',
        Default: -1,
        Help: ''
    },
    "CacheMode": {
        Type: 'options',
        Options: [
            {value:'off', key: 0},
            {value:'minimal', key: 1},
            {value:'writes', key: 2},
            {value:'full', key: 3},
        ],
        Default: 0,
        Help: ''
    },
    "CachePollInterval": {
        Type: 'int',
        Default: 60000000000,
        Help: ''
    },
    "CaseInsensitive": {
        Type: 'bool',
        Default: false,
        Help: ''
    },
    "ChunkSize":{
        Type: 'int',
        Default: 134217728,
        Help: ''
    },
    "ChunkSizeLimit": {
        Type: 'int',
        Default: -1,
        Help: ''
    },
    "DirCacheTime": {
        Type: 'int',
        Default: 300000000000,
        Help: ''
    },
    "DirPerms": {
        Type: 'int',
        Default: 511,
        Help: ''
    },
    "FilePerms": {
        Type: 'int',
        Default: 438,
        Help: ''
    },
    "GID": {
        Type: 'int',
        Default: 1000,
        Help: ''
    },
    "NoChecksum": {
        Type: 'bool',
        Default: false,
        Help: ''
    },
    "NoModTime": {
        Type: 'bool',
        Default: false,
        Help: ''
    },
    "NoSeek": {
        Type: 'bool',
        Default: false,
        Help: ''
    },
    "PollInterval": {
        Type: 'int',
        Default: 60000000000,
        Help: ''
    },
    "ReadAhead": {
        Type: 'int',
        Default: 0,
        Help: ''
    },
    "ReadOnly": {
        Type: 'bool',
        Default: false,
        Help: ''
    },
    "ReadWait": {
        Type: 'int',
        Default: 20000000,
        Help: ''
    },
    "UID": {
        Type: 'int',
        Default: 1000,
        Help: ''
    },
    "Umask": {
        Type: 'int',
        Default: 2,
        Help: ''
    },
    "WriteBack": {
        Type: 'int',
        Default: 5000000000,
        Help: ''
    },
    "WriteWait": {
        Type: 'int',
        Default: 1000000000,
        Help: ''
    }
}

export const mountOptions = {
    "AllowNonEmpty": {
        Type: 'bool',
        Default: false,
        Help: ''
    },
    "AllowOther": {
        Type: 'bool',
        Default: false,
        Help: ''
    },
    "AllowRoot": {
        Type: 'bool',
        Default: false,
        Help: ''
    },
    "AsyncRead": {
        Type: 'bool',
        Default: true,
        Help: ''
    },
    "AttrTimeout": {
        Type: 'int',
        Default: 1000000000,
        Help: ''
    },
    "Daemon": {
        Type: 'bool',
        Default: false,
        Help: ''
    },
    "DaemonTimeout": {
        Type: 'int',
        Default: 0,
        Help: ''
    },
    "DebugFUSE": {
        Type: 'bool',
        Default: false,
        Help: ''
    },
    "DefaultPermissions": {
        Type: 'bool',
        Default: false,
        Help: ''
    },
    "ExtraFlags": {
        Type: 'array',
        Default: [],
        Help: ''
    },
    "ExtraOptions": {
        Type: 'array',
        Default: [],
        Help: ''
    },
    "MaxReadAhead": {
        Type: 'int',
        Default: 131072,
        Help: ''
    },
    "NoAppleDouble": {
        Type: 'bool',
        Default: true,
        Help: ''
    },
    "NoAppleXattr": {
        Type: 'bool',
        Default: false,
        Help: ''
    },
    "VolumeName": {
        Type: 'string',
        Default: "",
        Help: ''
    },
    "WritebackCache": {
        Type: 'bool',
        Default: false,
        Help: ''
    }
}