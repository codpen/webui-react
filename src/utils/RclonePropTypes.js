import * as PropTypes from "prop-types";

export const PROP_FS_INFO = PropTypes.oneOfType([
    PropTypes.shape({
        Features: PropTypes.object.isRequired,
        Hashes: PropTypes.array.isRequired,
        Name: PropTypes.string.isRequired,
        Precision: PropTypes.number.isRequired,
        String: PropTypes.string.isRequired
    }),
    PropTypes.object
]);

export const PROP_CURRENT_PATH = PropTypes.shape({
    remoteName: PropTypes.string.isRequired,
    remotePath: PropTypes.string.isRequired
});

export const PROP_CONTAINER_ID = PropTypes.string;

export const PROP_LOAD_IMAGES_BOOL = PropTypes.bool;

export const PROP_IN_VIEWPORT = PropTypes.bool;

export const PROP_ITEM = PropTypes.shape({

    Path: PropTypes.string.isRequired,
    Name: PropTypes.string.isRequired,
    Size: PropTypes.number,
    MimeType: PropTypes.string,
    ModTime: PropTypes.string.isRequired,
    IsDir: PropTypes.bool.isRequired

});

export const PROP_BANDWIDTH = PropTypes.shape({
    // bytesPerSecond: PropTypes.number.isRequired,
    rate: PropTypes.string.isRequired

});
