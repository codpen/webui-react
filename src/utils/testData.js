export const TEST_REDUX_PROPS = {
    status: {
        isConnected: true,
        jobs: {
            bytes: 0,
            checks: 0,
            deletes: 0,
            elapsedTime: 737.732988329,
            errors: 0,
            fatalError: false,
            retryError: false,
            speed: 0,
            transfers: 0
        },
        speed: [
            {
                elapsedTime: 737.732988329,
                speed: 0,
                speedAvg: 0
            }
        ],
        runningAvgSpeed: 0,
        checkStatus: false,
        bandwidth: {
            bytesPerSecond: -1,
            rate: 'off'
        }
    },
    config: {
        providers: [{
            Name: 'b2',
            Description: 'Backblaze B2',
            Prefix: 'b2',
            Options: [
                {
                    Name: 'account',
                    Help: 'Account ID or Application Key ID',
                    Provider: '',
                    Default: '',
                    Value: null,
                    ShortOpt: '',
                    Hide: 0,
                    Required: true,
                    IsPassword: false,
                    NoPrefix: false,
                    Advanced: false,
                    DefaultStr: '',
                    ValueStr: '',
                    Type: 'string'
                },
                {
                    Name: 'key',
                    Help: 'Application Key',
                    Provider: '',
                    Default: '',
                    Value: null,
                    ShortOpt: '',
                    Hide: 0,
                    Required: true,
                    IsPassword: false,
                    NoPrefix: false,
                    Advanced: false,
                    DefaultStr: '',
                    ValueStr: '',
                    Type: 'string'
                },
                {
                    Name: 'endpoint',
                    Help: 'Endpoint for the service.\nLeave blank normally.',
                    Provider: '',
                    Default: '',
                    Value: null,
                    ShortOpt: '',
                    Hide: 0,
                    Required: false,
                    IsPassword: false,
                    NoPrefix: false,
                    Advanced: true,
                    DefaultStr: '',
                    ValueStr: '',
                    Type: 'string'
                },
                {
                    Name: 'test_mode',
                    Help: 'A flag string for X-Bz-Test-Mode header for debugging.\n\nThis is for debugging purposes only. Setting it to one of the strings\nbelow will cause b2 to return specific errors:\n\n  * "fail_some_uploads"\n  * "expire_some_account_authorization_tokens"\n  * "force_cap_exceeded"\n\nThese will be set in the "X-Bz-Test-Mode" header which is documented\nin the [b2 integrations checklist](https://www.backblaze.com/b2/docs/integration_checklist.html).',
                    Provider: '',
                    Default: '',
                    Value: null,
                    ShortOpt: '',
                    Hide: 2,
                    Required: false,
                    IsPassword: false,
                    NoPrefix: false,
                    Advanced: true,
                    DefaultStr: '',
                    ValueStr: '',
                    Type: 'string'
                },
                {
                    Name: 'versions',
                    Help: 'Include old versions in directory listings.\nNote that when using this no file write operations are permitted,\nso you can\'t upload files or delete them.',
                    Provider: '',
                    Default: false,
                    Value: null,
                    ShortOpt: '',
                    Hide: 0,
                    Required: false,
                    IsPassword: false,
                    NoPrefix: false,
                    Advanced: true,
                    DefaultStr: 'false',
                    ValueStr: 'false',
                    Type: 'bool'
                },
                {
                    Name: 'hard_delete',
                    Help: 'Permanently delete files on remote removal, otherwise hide files.',
                    Provider: '',
                    Default: false,
                    Value: null,
                    ShortOpt: '',
                    Hide: 0,
                    Required: false,
                    IsPassword: false,
                    NoPrefix: false,
                    Advanced: false,
                    DefaultStr: 'false',
                    ValueStr: 'false',
                    Type: 'bool'
                },
                {
                    Name: 'upload_cutoff',
                    Help: 'Cutoff for switching to chunked upload.\n\nFiles above this size will be uploaded in chunks of "--b2-chunk-size".\n\nThis value should be set no larger than 4.657GiB (== 5GB).',
                    Provider: '',
                    Default: 209715200,
                    Value: null,
                    ShortOpt: '',
                    Hide: 0,
                    Required: false,
                    IsPassword: false,
                    NoPrefix: false,
                    Advanced: true,
                    DefaultStr: '200M',
                    ValueStr: '200M',
                    Type: 'SizeSuffix'
                },
                {
                    Name: 'chunk_size',
                    Help: 'Upload chunk size. Must fit in memory.\n\nWhen uploading large files, chunk the file into this size.  Note that\nthese chunks are buffered in memory and there might a maximum of\n"--transfers" chunks in progress at once.  5,000,000 Bytes is the\nminimum size.',
                    Provider: '',
                    Default: 100663296,
                    Value: null,
                    ShortOpt: '',
                    Hide: 0,
                    Required: false,
                    IsPassword: false,
                    NoPrefix: false,
                    Advanced: true,
                    DefaultStr: '96M',
                    ValueStr: '96M',
                    Type: 'SizeSuffix'
                },
                {
                    Name: 'disable_checksum',
                    Help: 'Disable checksums for large (> upload cutoff) files',
                    Provider: '',
                    Default: false,
                    Value: null,
                    ShortOpt: '',
                    Hide: 0,
                    Required: false,
                    IsPassword: false,
                    NoPrefix: false,
                    Advanced: true,
                    DefaultStr: 'false',
                    ValueStr: 'false',
                    Type: 'bool'
                },
                {
                    Name: 'download_url',
                    Help: 'Custom endpoint for downloads.\n\nThis is usually set to a Cloudflare CDN URL as Backblaze offers\nfree egress for data downloaded through the Cloudflare network.\nThis is probably only useful for a public bucket.\nLeave blank if you want to use the endpoint provided by Backblaze.',
                    Provider: '',
                    Default: '',
                    Value: null,
                    ShortOpt: '',
                    Hide: 0,
                    Required: false,
                    IsPassword: false,
                    NoPrefix: false,
                    Advanced: true,
                    DefaultStr: '',
                    ValueStr: '',
                    Type: 'string'
                },
                {
                    Name: 'download_auth_duration',
                    Help: 'Time before the authorization token will expire in s or suffix ms|s|m|h|d.\n\nThe duration before the download authorization token will expire.\nThe minimum value is 1 second. The maximum value is one week.',
                    Provider: '',
                    Default: 604800000000000,
                    Value: null,
                    ShortOpt: '',
                    Hide: 0,
                    Required: false,
                    IsPassword: false,
                    NoPrefix: false,
                    Advanced: true,
                    DefaultStr: '1w',
                    ValueStr: '1w',
                    Type: 'Duration'
                }
            ]
        }],
        configDump: {
            local: {
                type: 'local'
            },
            mydrive: {
                token: '',
                type: 'drive'
            }
        },
        hasError: false
    },
    remote: {
        configs: {
            local: {
                Features: {
                    About: true,
                    BucketBased: false,
                    CanHaveEmptyDirectories: true,
                    CaseInsensitive: false,
                    ChangeNotify: false,
                    CleanUp: false,
                    Copy: false,
                    DirCacheFlush: false,
                    DirMove: true,
                    DuplicateFiles: false,
                    GetTier: false,
                    ListR: false,
                    MergeDirs: false,
                    Move: true,
                    OpenWriterAt: true,
                    PublicLink: false,
                    Purge: true,
                    PutStream: true,
                    PutUnchecked: false,
                    ReadMimeType: false,
                    ServerSideAcrossConfigs: false,
                    SetTier: false,
                    SetWrapper: false,
                    UnWrap: false,
                    WrapFs: false,
                    WriteMimeType: false
                },
                Hashes: [
                    'MD5',
                    'SHA-1',
                    'DropboxHash',
                    'QuickXorHash',
                    'Whirlpool'
                ],
                Name: 'local',
                Precision: 1,
                Root: '/home/negative0/Projects/rclone/rclone',
                String: 'Local file system at /home/negative0/Projects/rclone/rclone'
            }
        },
        remotes: [
            'local',
            'mydrive'
        ],
        files: {
            'local-': {
                time: '2019-07-13T07:43:19.196Z',
                files: [
                    {
                        Path: '.appveyor.yml',
                        Name: '.appveyor.yml',
                        Size: 1434,
                        MimeType: 'application/octet-stream',
                        ModTime: '2019-06-07T22:44:37.357539000+05:30',
                        IsDir: false
                    },
                    {
                        Path: '.circleci',
                        Name: '.circleci',
                        Size: -1,
                        MimeType: 'inode/directory',
                        ModTime: '2019-06-24T09:45:29.483620000+05:30',
                        IsDir: true
                    }
                ]
            }
        },
        hasError: false,
        numCols: 1,
        distractionFreeMode: false
    },
    explorer: {
        backStacks: {
            '0': {
                backStack: {
                    items: [
                        {
                            remoteName: 'local',
                            remotePath: ''
                        }
                    ],
                    count: 1
                },
                forwardStack: {
                    items: [],
                    count: 0
                }
            }
        },
        currentPaths: {
            '0': {
                remoteName: 'local',
                remotePath: ''
            }
        },
        visibilityFilters: {},
        gridMode: {},
        searchQueries: {
            '0': ''
        },
        loadImages: {
            '0': false
        }
    },
    providerStatus: {
        about: {}
    },
    user: {
        auth: {
            userName: 'chaitanya',
            password: 'abcd',
            ipAddress: 'http://localhost:5572/',
            interceptor: 2
        }
    },
    imageLoader: {}
};

export const TEST_FILE_CONTAINER_ID = '0'

export const FILE_ITEM_DATA = {
    Path: '.appveyor.yml',
    Name: '.appveyor.yml',
    Size: 1434,
    MimeType: 'application/octet-stream',
    ModTime: '2019-06-07T22:44:37.357539000+05:30',
    IsDir: false
};

export const DIR_ITEM_DATA = {
    Path: 'abc',
    Name: 'abc',
    Size: 0,
    MimeType: '',
    ModTime: '2019-06-07T22:44:37.357539000+05:30',
    IsDir: true
};