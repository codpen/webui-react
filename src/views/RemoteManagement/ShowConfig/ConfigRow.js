import React from "react";
import axiosInstance from "../../../utils/API/API";
import {Button} from "reactstrap";
import * as  PropTypes from "prop-types";
import {toast} from "react-toastify";
import {withRouter} from "react-router-dom";


class ConfigRow extends React.Component {
    constructor(props, context) {
        super(props, context);
        let {remote, remoteName} = this.props;
        remote["name"] = remoteName;
        this.state = {
            remote: remote
        };
        this.onDeleteClicked = this.onDeleteClicked.bind(this);
        this.onUpdateClicked = this.onUpdateClicked.bind(this);

    }


    onUpdateClicked = () => {
        const {name} = this.state.remote;
        this.props.history.push("/newdrive/edit/" + name);
    };

    // TODO: Delete config functionality
    onDeleteClicked() {
        const {name} = this.state.remote;
        let {refreshHandle} = this.props;

        // Delete http request
        if (window.confirm(`Are you sure you wish to delete ${name}? You cannot restore it once it is deleted.`)) {

            axiosInstance.post("/config/delete", {name: name}).then(
                (res) => {
                    // console.log(res);
                    // Refresh the parent component
                    refreshHandle();
                    toast.info('Config deleted');
                }, (err) => {
                    // console.log(`Error occurred: ${err}`);
                    toast.error('Error deleting config')
                }
            )
        }
    }


    render() {
        const {name, type} = this.state.remote;
        const {sequenceNumber} = this.props;
        return (
            <tr data-test="configRowComponent">
                <th scope="row">{sequenceNumber}</th>
                <td>{name}</td>
                <td>{type}</td>
                <td>

                    <Button className={"bg-info"} onClick={this.onUpdateClicked}>Update</Button>
                </td>
                <td><Button className={"bg-danger"} onClick={this.onDeleteClicked}>Delete</Button></td>
            </tr>
        );
    }
}

const propTypes = {
    remote: PropTypes.object.isRequired, // Name of the remote to perform operations
    refreshHandle: PropTypes.func.isRequired, // Used to refresh the parent component upon change
    sequenceNumber: PropTypes.number.isRequired,
    remoteName: PropTypes.string.isRequired,

};

ConfigRow.propTypes = propTypes;

export default withRouter(ConfigRow);