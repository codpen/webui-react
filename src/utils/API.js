import axios from "axios";
import './Global'

let axiosInstance = axios.create({
    baseURL: global.ipAddress,
    headers: {'Content-Type': 'application/json'},
    responseType: "json"
});

export function performMoveFile(srcFs, srcRemote, dstFs, dstRemote, Name, IsDir) {
    return performCopyOrMoveFile(srcFs, srcRemote, dstFs, dstRemote, Name, IsDir, 'move');
}

export function performCopyFile(srcFs, srcRemote, dstFs, dstRemote, Name, IsDir) {

    return performCopyOrMoveFile(srcFs, srcRemote, dstFs, dstRemote, Name, IsDir, 'copy');
}

async function performCopyOrMoveFile(srcFs, srcRemote, dstFs, dstRemote, Name, IsDir, mode) {
    let url = "";
    if (mode === "move") {
        url = "operations/movefile";
    } else {
        url = "/operations/copyfile";
    }
    if (IsDir) {
        // if (dstRemote === "") {
        //     dstRemote = Name;
        // } else {
        //     dstRemote += "/" + Name;
        // }
        // const data = {
        //     srcFs: srcFs + srcRemote,
        //     dstFs: dstFs + dstRemote
        // };

        // let res = await axiosInstance.post("/operations/copy")
        // let res = await axiosInstance.post("/sync/copy", data);
        // console.log("Dir copy Res", res);
        // this.setState({isOperationInProgress: false})
        console.log("Copying directory feature is not yet implemented");

    } else {
        if (dstRemote === "") {
            dstRemote = Name;
        } else {
            dstRemote += "/" + Name;
        }

        const data = {
            srcFs: srcFs,
            srcRemote: srcRemote,
            dstFs: dstFs,
            dstRemote: dstRemote,
        };
        let res = await axiosInstance.post(url, data);
        console.log("Res", res);

    }
}

export default axiosInstance;
