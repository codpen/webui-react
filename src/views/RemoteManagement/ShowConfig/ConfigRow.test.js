import React from "react";
import {shallow} from "enzyme";
import {findByTestAttr, testStore} from "../../../../Utils";
import toJson from "enzyme-to-json";
import {TEST_REDUX_PROPS} from "../../../utils/testData";
import ConfigRow from "./ConfigRow";

const setUp = (intialState = {}, props = {}) => {
    const store = testStore(intialState);
    // Wrapped component is required to bypass the withRouter HOC and render only the component
    const component = shallow(<ConfigRow.WrappedComponent {...props} store={store}/>);
    return component;
};


describe('Config Row', function () {

    let refreshHandle = jest.fn();

    describe('renders', function () {
        let wrapper;
        beforeEach(() => {
            const initialState = {
                ...TEST_REDUX_PROPS
            };
            const props = {
                sequenceNumber: 1,
                key: 1,
                remoteName: 'local',
                remote: {
                    token: '',
                    type: 'drive'
                },
                refreshHandle: refreshHandle
            };
            wrapper = setUp(initialState, props)
        });

        it('should render without crashing', function () {
            const component = findByTestAttr(wrapper, "configRowComponent");
            expect(component).toHaveLength(1);
        });

        it('should match snapshot', function () {
            expect(toJson(wrapper)).toMatchSnapshot()
        });

    });

});