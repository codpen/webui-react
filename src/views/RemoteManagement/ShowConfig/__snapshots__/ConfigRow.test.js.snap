// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Row renders should match snapshot 1`] = `
<tr
  data-test="configRowComponent"
>
  <th
    scope="row"
  >
    1
  </th>
  <td>
    local
  </td>
  <td>
    drive
  </td>
  <td>
    <Button
      className="bg-info mr-2"
      color="secondary"
      onClick={[Function]}
      tag="button"
    >
      Update
    </Button>
    <Button
      className="bg-danger"
      color="secondary"
      onClick={[Function]}
      tag="button"
    >
      Delete
    </Button>
  </td>
</tr>
`;
